pkg/services/tree/redirect.go

```diff
@@ -0,0 +1,65 @@
+package tree
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/nspcc-dev/neo-go/pkg/crypto/keys"
+	"github.com/nspcc-dev/neofs-node/pkg/network"
+	cidSDK "github.com/nspcc-dev/neofs-sdk-go/container/id"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+)
+
+var errNoSuitableNode = errors.New("no node was found to execute the request")
+
+// forEachNode executes callback for each node in the container.
+// If the node belongs to a container, nil error is returned.
+// Otherwise, f is executed for each node, stopping if true is returned.
+func (s *Service) forEachNode(ctx context.Context, cid *cidSDK.ID, f func(c TreeServiceClient) bool) error {
+	cntNodes, err := s.getContainerNodes(cid)
+	if err != nil {
+		return fmt.Errorf("can't get container nodes: %w", err)
+	}
+
+	rawPub := (*keys.PublicKey)(&s.key.PublicKey).Bytes()
+	for _, node := range cntNodes.Flatten() {
+		if bytes.Equal(node.PublicKey(), rawPub) {
+			return nil
+		}
+	}
+
+	nodes := cntNodes.Flatten()
+	for _, node := range nodes {
+		var stop bool
+		node.IterateAddresses(func(endpoint string) bool {
+			c, err := dialTreeService(ctx, endpoint)
+			if err != nil {
+				return false
+			}
+
+			s.log.Debug("redirecting tree service query", zap.String("endpoint", endpoint))
+			stop = f(c)
+			return true
+		})
+		if stop {
+			return nil
+		}
+	}
+	return errNoSuitableNode
```

Comment:
did i get it right: if i pass a callback that always returns `false` i get `errNoSuitableNode` err as a result even if there are available nodes in the container? was it expected?

Reply:
Nice catch, fixed.